### PR TITLE
Add megabits/s and gigabits/s as units of Rate

### LINKED ISSUE #1065: CCB-91: Add more units of Rate: megabits/s and gigabits/s
### 🦄 Sibling SCR

NASA-PDS/PDS4-CCB#91

### 💡 SCR Summary

Add new units of rate:

megabits/s
gigabits/s

### Added to parent epic?

Will do

### PR DESCRIPTION
**Pull request by jshughes**
Add more units of Rate: megabits/s and gigabits/s

## 🗒️ Summary
Added two new units of rate, megabits/s and gigabits/s, to unit_id in Units_of_Rates

### 🤖 AI Assistance Disclosure

- [x] AI used for light assistance (e.g., suggestions, refactoring, documentation help, minor edits)

Estimated % of code influenced by AI: _10_ %
Identify and debug Cucumber tests.

## ⚙️ Test Data and/or Report
github1065 - Added four validate tests: one control, two valid, and one invalid. 

## ♻️ Related Issues
fixes #1065
refs NASA-PDS/PDs4-CCB#91 

## 🤓 Reviewer Checklist

*Reviewers: Please verify the following before approving this pull request.*

### Documentation and PR Content
- [ ] **Documentation:** README, Wiki, or inline documentation (Sphinx, Javadoc, Docstrings) have been updated to reflect these changes.
- [ ] **Issue Traceability:** The PR is linked to a valid GitHub Issue
- [ ] **PR Title:** The PR title is "user-friendly" clearly identifying what is being fixed or the new feature being added, that if you saw it in the Release Notes for a tool, you would be able to get the gist of what was done.

### Security & Quality
- [ ] **SonarCloud:** Confirmed no new High or Critical security findings.
- [ ] **Secrets Detection:** Verified that the Secrets Detection scan passed and no sensitive information (keys, tokens, PII) is exposed.
- [ ] **Code Quality:** Code follows organization style guidelines and best practices for the specific language (e.g., PEP 8, Google Java Style).

### Testing & Validation
- [ ] **Test Accuracy:** Verified that test data is accurate, representative of real-world PDS4 scenarios, and sufficient for the logic being tested.
- [ ] **Coverage:** Automated tests cover new logic and edge cases.
- [ ] **Local Verification:** (If applicable) Successfully built and ran the changes in a local or staging environment.

### Maintenance
- [ ] **Backward Compatibility:** Confirmed that these changes do not break existing downstream dependencies or API contracts (or that breaking changes are clearly documented).

### Terraform (only if this PR touches a `terraform/` directory)

- [ ] **Validator run:** `scripts/validate_terraform.py` was run against the changed `terraform/` tree and any Must-Have failures were resolved (or are tracked in a reviewed `.tfvalidate-ignore` entry, not silently bypassed).
- [ ] **Ownership boundary:** Shared/singleton CDS resources (Cognito, CloudFront, org-wide IAM roles, shared security groups) were only added/changed in `pdc-cds-infra`, not duplicated in an application repo.
- [ ] **Module structure:** `modules/` stays flat and no new module is just a thin wrapper around a single resource.
- [ ] **SSM interface:** Any new cross-component output is actually published under `/pds/<component>/...` and, if this PR is the consumer side, reads from the correct published parameter rather than another repo's Terraform state.
- [ ] **Naming intent:** Resource and variable names are semantically meaningful, not just non-redundant (the validator only checks for repeated resource-type substrings).
- [ ] **Auth at runtime:** Terraform/CI actually authenticates via OIDC or an assumed role when applied — not just "no static key found in this diff."
- [ ] **Least privilege:** Any new or changed IAM policy attached to the Terraform execution role grants only what this change needs.

### CODE REVIEW

**Review by jmafi — APPROVED** [review]
I'm not sure what many of these files are (".pins", ".feature", ".pont", ".pprj"), but the diffs on them look to be in line with requested changes on CCB#91. The test labels also look good.

**Comment by jordanpadams** [comment]
@jshughes

> I'm not sure what many of these files are (".pins", ".feature", ".pont", ".pprj"), but the diffs on them look to be in line with requested changes on CCB#91. The test labels also look good.

.feature files are the test config files for Cucumber testing.

the rest I think are outputs from Protege? does your software need those? if not, we should look to clean those out.

**Comment by jordanpadams** [comment]
@rchenatjpl can you review these test cases and verify they are sufficient? the test expectation is here: https://github.com/NASA-PDS/pds4-information-model/pull/1066/changes#diff-ea476d3e7ba5802b1744479a8cec266285af4fe081e8213156e6cdd85495f4baR92

File: model-lddtool/src/test/resources/features/validate.feature [hunk 1]
```diff
@@ -85,3 +85,9 @@ Feature: pds4_information_model_validate_integration
...
+
+    @v18.0.x
+    Examples: 
+| testId                                | testName                                                               | testDir       | messageCount| problemEnum         | commandArgs                                                                                                                                                                                                                                                                                 | ingestLDDFileName                            | pds4Version | enableContentValidation |
+| NASA-PDS/pds4-information-model#1065  | "New units megabits/s and gigabits/s to Units_of_Rates"                | "github1065"  |          3  | "totalErrors"       | "-t {resourceUpdateVersionDir}/github1065/tc_Control_Valid_Unit_kilobits.xml {resourceUpdateVersionDir}/github1065/tc_New_Valid_Unit_megabits.xml {resourceUpdateVersionDir}/github1065/tc_New_Valid_Unit_gigabits.xml {resourceUpdateVersionDir}/github1065/tc_Invalid_Unit_gigabytes.xml" | "PDS4_PARTICLE_IngestLDD_UnitsOfMeasure.xml" | "" | "false" |
```
Comment (jordanpadams) [inline model-lddtool/src/test/resources/features/validate.feature:92]: @jshughes when I generated the schemas and run against these files, only 1 should fail, no? why is the test expectation here for 3 errors? that would mean this fix actually didn't work.

File: model-lddtool/src/test/resources/features/validate.feature [hunk 2]
```diff
@@ -85,3 +85,9 @@ Feature: pds4_information_model_validate_integration
 | NASA-PDS/pds4-information-model#1043d | "reconciled namespaces"                                                | "github1043d" |           0 | "totalErrors"     | "-t {resourceUpdateVersionDir}/github1043d/namespace_mpf_Pass_260618.xml"     | "PDS4_PARTICLE_IngestLDD_mpf.xml"     | "" | "false" |
 | NASA-PDS/pds4-information-model#1043e | "reconciled namespaces"                                                | "github1043e" |           0 | "totalErrors"     | "-t {resourceUpdateVersionDir}/github1043e/namespace_mvn_Pass_260618.xml"     | "PDS4_PARTICLE_IngestLDD_mvn.xml"     | "" | "false" |
 | NASA-PDS/pds4-information-model#1043f | "reconciled namespaces"                                                | "github1043f" |           0 | "totalErrors"     | "-t {resourceUpdateVersionDir}/github1043f/namespace_orex_Pass_260618.xml"    | "PDS4_PARTICLE_IngestLDD_orex.xml"    | "" | "false" |
+
+    @v16.2.x
```
Comment (jordanpadams) [inline model-lddtool/src/test/resources/features/validate.feature:89]: @jshughes note: I updated this version here. These software versions are not tied to build numbers. Builds are arbitrary incrementing counters. our software uses semantic versioning

**Review by jordanpadams — CHANGES_REQUESTED** [review]

**Comment by jshughes** [comment]
@jordanpadams @rchenatjpl @jmafi The .pins, .pont, and .pprj files together comprise a Protégé database. The Information Model specification consists of two Protege databases: UpperModel and dd11179. The .pprj file contains high-level project information and is not used by the application; however, it is included to preserve the completeness and consistency of each database. However, it could be removed if preferred.

File: model-lddtool/src/test/resources/features/validate.feature (hunk 1 above)
Comment (jordanpadams) [reply model-lddtool/src/test/resources/features/validate.feature:92]: This may be a bug in the test code, which may require some AI help to figure out.

Comment (jshughes) [reply model-lddtool/src/test/resources/features/validate.feature:92]: Both Oxygen and the Cucumber tests raise three errors for the one invalid case. The module might be the Xerces XSD processor. 

An AI prompt to explain the three Oxygen errors resulted in the following:
Oxygen XML Editor performed three separate operations on the test file. The results are generally consistent with a deliberately invalid test case named tc_Invalid_Unit_gigabytes.xml.
1.	No CSS files available
2.	Content Completion schemas changed
3.	The OPS Schematron detected one OPS-specific rule violation

The first two messages are informational notices, not XML validation errors.

There are some suggested solutions for the first error, such as use Text or Grid mode instead of Author mode in Oxygen. Will need to research how this affects the Cucumber test. Still trying to understand the schema changed error.

File: model-lddtool/src/test/resources/features/validate.feature (hunk 2 above)
Comment (jshughes) [reply model-lddtool/src/test/resources/features/validate.feature:89]: Thanks for the correction.

Comment (rchenatjpl) [reply model-lddtool/src/test/resources/features/validate.feature:89]: @jordanpadams Hi, I apologize, but how urgent is this? It's going to take me half a day to figure out / remember how to get started with this.

**Comment by jordanpadams** [comment]
@rchenatjpl if you could review and identify any issues by beginning of next week that would be great. without a merge this often blocks Steve from proceeding with new work.

you really just need to:
1. check the SCR for what was supposed to be done
2. track down the label files (they are in this repo)
3. validate them to see if they work. for this particular test case, you will need to also generate an LDD (IngestLDD also in this repo).

TLDR; we need people other than me to know what is going on here and who can test this because it is not scalable for me to continue to be the blocker here.

**Comment by jshughes** [comment]
Oxygen reports three validation problems for the invalid unit, but they
represent one underlying data error.

	The XML uses an invalid unit value—apparently gigabytes—where the applicable
	PDS4 definition permits only specified units such as kilobits. Oxygen detects
	this single defect through two independent validation mechanisms:

		1) XML Schema validation produces two diagnostics. These are likely separate
		consequences of the same invalid value, such as failure against the controlled-
		value restriction followed by failure to match the expected element/type
		definition.

		2) OPS Schematron validation produces one additional diagnostic because the same
		value violates an OPS business rule.

**Comment by jshughes** [comment]
@jordanpadams I was hoping that there was some type of name collision with the test attribute name, "unit_type". That is not the problem.

F1) The Claude code testing suggested removing the schematron rule in the PDS4_OPS_1R00_1010.sch file. That reduced the error count to 2. 

F2) Since the LDD schema, PDS4_OPS_1R00_1010.xsd, references the core schema, PDS4_PDS_1R00.xsd, for the definition of Units_of_Rates, it seems that the validation error raised by PDS4_PDS_1R00.xsd causes the same error to be raised by PDS4_OPS_1R00_1010.xsd.

F1 seems to suggest that LDDTool should be modified to not write a rule for any unit_of_measure_type to an LDD schematron file.

F2 seems to suggest that we will get an error from each schema involved in a validation.

Your thoughts?

**Comment by jshughes** [comment]
@jordanpadams I submitted F2 to Claude code for further analysis. The summary is: Unchanged from the earlier finding, and now on firmer ground: the 2 XSD messages cannot be reduced by moving, inlining, or re-homing Units_of_Rates — the pair is Xerces behavior for any failed enumeration facet.

**Comment by jordanpadams** [comment]
@jshughes I agree all of the units checks should probably be removed from schematron, which would be interesting to see how that could improve performance, but I think that is out of scope for this task. The current implementation is successful. Feel free to create a new ticket for that refactor, which will require a bit more testing.

**Review by jordanpadams — APPROVED** [review]

**Comment by rchenatjpl** [comment]
@jordanpadams @jshughes days late, dollars short: I got three errors as well
```tc_Invalid_Unit_gigabytes.xml (urn:nasa:pds:context:investigation:mission.dummy::1.0)
      ERROR  [error.label.schema]   line 21, 39: cvc-enumeration-valid: Value 'gigabytes' is not facet-valid with respect to enumeration '[counts/bin, day**-1, gigabits/s, kilobits/s, megabits/s]'. It must be a value from the enumeration.
      ERROR  [error.label.schema]   line 21, 39: cvc-attribute.3: The value 'gigabytes' of attribute 'unit' on element 'ops:unit_type' is not valid with respect to its type, 'Units_of_Rates'.
      ERROR  [error.label.schematron]   line 21, 39: The attribute @unit must be equal to one of the following values 'counts/bin', 'day**-1', 'gigabits/s', 'kilobits/s', 'megabits/s'.
```
What is the ops: dictionary? I thought it would be the one for our resources, but the contents don't look like it. In any case, if I were starting fresh, I'd create a dummy Product_Observational that uses it, but I don't think they'd be better than Steve's Product_Context samples.

**Comment by jshughes** [comment]
Regarding ops:, the smallish PDS4_PARTICLE_IngestLDD.xml was copied and simplified down to one attribute. The plan is to re-use it for testing future values of units of measure.

**Comment by jshughes** [comment]
The ticket #1067 has been opened to look into removing the schematron rule. This would reduce the error count to 2.
